### ISSUE TITLE
Implement rule for compound functions in GAF pipeline

### ISSUE THREAD

**Issue by cmungall** [biolink/ontobio#341]
We want to implement the axiom `enables o has_part -> enables` as a pre-inference in the GOC annotation pipeline. This rule is for inferring subfunction annotations from compound functions.

The background is here: https://github.com/geneontology/go-ontology/issues/16972 (use the go-ontology ticket to comment on the biology or axioms, this ticket to discuss the implementation)

We will first need to make a go-rule. I am doing this as an ontobio ticket first as it's easier for me and I'm lazy. Also we want to do experiments before making the rule.

Note: if we use an OWL reasoner as part of the pipeline then this is all taken care of by axioms in the ontology, but for now it's easier to write as an ad-hoc rule, as follows.

 * For any MF annotation between `G` and `CF` that has no qualifier (i.e. the semantics are `enables`)
     * where `CF-anc` is the ancestor of F over reflexive is_a closure (i.e. `reflexive=True, relations=['subClassOf']` in ontobio)
     * if there is an direct edge `CF-anc has-part SF` in the ontology then make another annotation between `G` and `SF`. Copy evidence etc, as we do for F->P inferences

@dougli1sqrd will implement this as a PR, we will run a text of the pipeline from that and can do a diff of annotatins for any group

**Comment by dougli1sqrd** [comment]
Let me write down what I think this means:

For an annotation:
`gp enables (mf subClassOf MF)`
Where gp is a gene product, MF is Molecular Function,
`mf subClassof mf-anc`
`mf-anc has_part (sf subClassOf MF)` Where sf means sub function
create a new annotation of `gp enables sf`

**Comment by cmungall** [comment]
correct

**Comment by dougli1sqrd** [comment]
So I have a beginning implementation of this. Do we have an example annotation that should yield inferences?

**Comment by dougli1sqrd** [comment]
So, I have pushed https://github.com/biolink/ontobio/tree/issue-341-enables_script, and you can find a new script: materialize.py in bin.
Run with:
`python bin/materialize.py infer --ontology go-ontology.json --gaf target/groups/mgi/mgi.gaf --target mgi_enables_inferred.gaf`
inside an installed ontobio and virtualenv. 

Please let me know how this is, and let me know what improvements I should make! This was a cool project. So far though, I feel skeptical that it's right. My MGI test hans't revealed any inference annotations yet. And it's slow! But at least we can collaborate on the code I have here.

**Comment by cmungall** [comment]
Can you make a WIP PR?

**Comment by cmungall** [comment]
Basic logic seems sound! You shouldn't use logical definitions. Recall that you will be operating over a released ontology that will have had `relax` executed. Just get has-part edges from ontobio in the usual way. This will likely give you inferences.

Regarding speed, of course this will be much faster after #343 is merged.

**Comment by cmungall** [comment]
I might generalize some of this into a method `get_inferred_restrictions` which takes a relation R and returns all inferred ancestors `R some X`.

**Comment by dougli1sqrd** [comment]
Yeah generalizing this would be easy!

**Comment by dougli1sqrd** [comment]
One thing is I'm not sure how to get the GO term taxon info out of the ontobio Ontology graph. Using the edges is definitely much faster.

**Comment by cmungall** [comment]
I don't follow - this should have nothing to do with taxa

**Comment by dougli1sqrd** [comment]
Okay, so the updated version is faster, using edges. Here is the small truncated mgi file I'm testing with: [mgi.smol.gaf.gz](https://github.com/biolink/ontobio/files/3263531/mgi.smol.gaf.gz), and the inferred output: [mgi_enables_inferred.gaf.gz](https://github.com/biolink/ontobio/files/3263538/mgi_enables_inferred.gaf.gz)

(the input mgi gaf should be unzipped first before trying)

I used a fairly recent go ontology json file. Does this look right to you?

**Comment by cmungall** [comment]
I'm not sure how you arrived at this

```
MGI     MGI:1917295     2610528A11Rik           GO:0044183      MGI:MGI:6094556|PMID:28900043   IDA             F       RIKEN cDNA 2610528A11 gene              protein taxon:10090     20171027        UniProt
M
```

given it is not the target of any has-parts (hacky check:`grep 'has_part GO:0044183' go-edit.obo` returns no lines)

Is this the result of all inferences, e.g including capable-of?

Is it possible to do a report similar to the ones owltools does?

**Comment by dougli1sqrd** [comment]
@cmungall For your previous comment:

> I don't follow - this should have nothing to do with taxa

Look at this line: https://github.com/biolink/ontobio/pull/347/files#diff-6e56ac2f5dec9a3fc80adc19641a0fbfR117

**Comment by pgaudet** [comment]
Files showing the inferences created by this rule are here: 
http://skyhook.berkeleybop.org/special_inferences/products/annotations/

**Comment by dougli1sqrd** [comment]
Updated files are in the skyhookd location

**Comment by dougli1sqrd** [comment]
I also merged the script used above. I'll close this unless there's something else.

**Comment by pgaudet** [comment]
@dougli1sqrd @cmungall In practice, what does this do ? The inferences are exported as part of the pipeline as 'predicted annotations'?

**Comment by dougli1sqrd** [comment]
Right now the results are just in that link above. This is not run in the pipeline. My understanding was this was a one-off script. Since the results have been produced and the script is merged, if we want specifically to put it in the pipeline maybe that should be a separate ticket?

**Comment by pgaudet** [comment]
The task described by Chris in the ticket is 'We want to implement the axiom enables o has_part -> enables as a pre-inference in the GOC annotation pipeline. ' 

So making the report as a one-off is only ONE step.

**Comment by cmungall** [comment]
@pgaudet is correct. @dougli1sqrd if you would like to split the work over two tickets in different repos that is OK (check with @kltm on what best practice is). But if you do then you should open up that ticket yourself and assign yourself and link back here.

**Comment by dougli1sqrd** [comment]
Cool, thanks @cmungall and @pgaudet! I will make a new ticket in `pipeline` to incorporate those changes into the pipeline then.
